docs/content/v0.20/architecture.md

```diff
@@ -1,36 +1,433 @@
+</figure>
+
+At startup, the controller code sets up the dynamic informers, the event
+handler and the work queue as follows:
+
+- lists all API resources
+- Filters out some resources
+- For each resource:
+  - Creates GVK key
+  - Registers Event Handler
+  - Starts Informer
+  - Indexes informer and lister in a map by GVK key
+- Waits for all caches to sync
+- Starts N workers to process work queue
+
+The reflector is started as part of the informer and watches specific
+resources on the WDS API Server; on create/update/delete object events it
+puts a copy of the object into the local cache. The informer invokes the
+event handler. The handler implements the event handling functions
+(AddFunc, UpdateFunc, DeleteFunc)
+
+A typical event flow for a create/update object event will run as
+follows:
+
+1.  Informer invokes the event handler AddFunc or UpdateFunc
+
+2.  The event handler does some filtering (for example, to ignore update
+    events where the object content is not modified) and then creates a
+    key to store a reference to the object in the work queue. The key
+    contains the GVK key used to retrieve a reference to the informer
+    and lister for that kind of object, and a namespace + name key to
+    retrieve the actual object. Storing the key in the work queue is a
+    common pattern in client-go as the object may have changed in the
+    cache (which is kept updated by the reflector) by the time a worker
+    gets a copy from the work queue. Workers should always receive the
+    key and use it to retrieve the object from the cache.
+
+3.  A worker pulls a key from the work queue, and then does the
+    following processing:
+
+    -  Uses the GVK key to get the reference to the lister for the
+        object
+    -  Gets the object from the lister cache using the NamespacedName of 
+       the object.
+    -  Gets the lister for Placement objects and list all placements
+    -  Iterates on all placements, and for each placement:
+        - evaluates whether the object matches the downsync selection 
+          criteria in the Placement.
+        - If there is a match, list ManagedClusters and
+          find the matching clusters using the label selector expression
+          for clusters.
+        - If there are matching clusters, add the names of the cluster
+          to a hashmap setting the name of the cluster as a key. Cluster
+          groups from different placements are merged together.
+        - If the one of the matched Placements has ` WantSingletonReportedState`
+          set to true, clusters are sorted in alphanumerical order and
+          only the first cluster is selected for delivery.  
+    -  If there are no matching clusters, the worker returns without
+        actions and is ready to process other events from the queue.
+    -  If there are matching clusters:
+        - Wraps the object into a ManifestWork
+        - Adds a label for each matched (Placement, WDS) to the ManifestWork that 
```

Comment:
Speaking of "each matched Placement" grates against the consideration for WantSingletonReportedState. Does having that bit set anywhere change the "each matched Placement" here to "the Placement selected for delivery"? And what about the Placements that matched but were not selected for delivery?

Reply:
Currently, setting `WantSingletonReportedState` in one of the placements that matches the object affects the behavior for all matching placements,  those selected for delivery (with matching clusters) and those not selected for delivery.

Reply:
The text that I am commenting on here says unequivocally that there is a label added for every matching Placement. Is this really true even if some of those matching Placements have `WantSingletonReportedState == true` ?  Also, the following bullet speaks simply of "each cluster". Is that really each cluster matching any Placement that matches the object, regardless of `WantSingletonReportedState` ? I doubt it. I think these pieces of text need updating. Is anything undone in the `WantSingletonReportedState == true` case for clusters that match a Placement that matches the object but are _not_ selected for delivery? The text should say.

Reply:
Since we agreed with @ezrasilvera and @MikeSpreitzer to move ahead and merge, I moved this discussion in this issue https://github.com/kubestellar/kubestellar/issues/1688